File: code/ex207.java
// Topological ordering

class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int n = prerequisites.length;
        
        // Get the indegree of each course
        int[] indegree = new int[numCourses];
        for (int[] pr : prerequisites) {
            indegree[pr[1]]++;
        }
        
        // Run topological ordering
        Queue<Integer> queue = new LinkedList<>();
        for (int i = 0; i < numCourses; i++) {
            if (indegree[i] == 0)    queue.offer(i);
        }
        if (queue.isEmpty())    return false;
        
        int visited = 0;
        while (!queue.isEmpty()) {
            int out = queue.poll();
            visited++;
            
            for (int i = 0; i < n; i++) {
                if (prerequisites[i][0] != out)    continue;
                
                if (--indegree[prerequisites[i][1]] == 0) {
                    queue.offer(prerequisites[i][1]);
                }
            }
        }
        
        return visited == numCourses;
    }
}

File: code/ex207.cpp
class Solution {
public:
    bool canFinish(int n, vector<vector<int>>& prerequisites) {
        vector<vector<int>> next(n, vector<int>());
        
        // Get indegree
        vector<int> indegree(n);
        for (auto& pre : prerequisites) {
            next[pre[1]].push_back(pre[0]);
            indegree[pre[0]]++;
        }
        
        int cnt = 0;
        queue<int> que;
        for (int i = 0; i < n; i++) {
            if (indegree[i] == 0) {
                que.push(i);
                cnt++;   
            }
        }
        
        while (!que.empty()) {
            int out = que.front();    que.pop();
            for (int nxt : next[out]) {
                if (--indegree[nxt] == 0) {
                    que.push(nxt);
                    cnt++;
                }
            }
        }
        
        return cnt == n;
    }
};